src/ccl_f1d.c

```diff
@@ -12,44 +12,168 @@
 //x     -> x-axis
 //y     -> f(x)-axis
 //y0,yf -> values of f(x) to use beyond the interpolation range
-ccl_f1d_t *ccl_f1d_t_new(int n,double *x,double *y,double y0,double yf)
+ccl_f1d_t *ccl_f1d_t_new(int n,double *x,double *y,double y0,double yf,
+			 ccl_f1d_extrap_t extrap_lo_type,
+			 ccl_f1d_extrap_t extrap_hi_type, int *status)
 {
   ccl_f1d_t *spl=malloc(sizeof(ccl_f1d_t));
-  if(spl==NULL)
+  if(spl==NULL) {
+    *status = CCL_ERROR_MEMORY;
     return NULL;
+  }
 
-  spl->spline=gsl_spline_alloc(gsl_interp_cspline,n);
+  spl->spline=gsl_spline_alloc(gsl_interp_akima,n);
   if (spl->spline == NULL) {
+    *status = CCL_ERROR_MEMORY;
     free(spl);
     return NULL;
   }
   int parstatus=gsl_spline_init(spl->spline,x,y,n);
   if(parstatus) {
+    *status = CCL_ERROR_SPLINE;
     gsl_spline_free(spl->spline);
     free(spl);
     return NULL;
   }
 
-  spl->x0=x[0];
-  spl->xf=x[n-1];
-  spl->y0=y0;
-  spl->yf=yf;
+  if(*status==0) {
+    spl->y0=y0;
+    spl->yf=yf;
+    spl->x_ini=x[0];
+    spl->x_end=x[n-1];
+    spl->y_ini=y[0];
+    spl->y_end=y[n-1];
+    spl->extrap_lo_type=extrap_lo_type;
+    spl->extrap_hi_type=extrap_hi_type;
+  }
+
+  // Compute derivatives
+  // Low-end
+  if(spl->extrap_lo_type == ccl_f1d_extrap_const) {
+    spl->der_lo = 0;
+  } 
+  else if(spl->extrap_lo_type == ccl_f1d_extrap_linx_liny) {
+    spl->der_lo = (y[1]-y[0])/(x[1]-x[0]);
+  }
+  else if(spl->extrap_lo_type == ccl_f1d_extrap_linx_logy) {
+    if(y[1]*y[0]<=0)
+      *status=CCL_ERROR_SPLINE;
+    else
+      spl->der_lo = log(y[1]/y[0])/(x[1]-x[0]);
+  }
+  else if(spl->extrap_lo_type == ccl_f1d_extrap_logx_liny) {
+    if(x[1]*x[0]<=0)
+      *status=CCL_ERROR_SPLINE;
+    else
+      spl->der_lo = (y[1]-y[0])/log(x[1]/x[0]);
+  }
+  else if(spl->extrap_lo_type == ccl_f1d_extrap_logx_logy) {
+    if((y[1]*y[0]<=0) || (x[1]*x[0]<=0))
+      *status=CCL_ERROR_SPLINE;
+    else
+      spl->der_lo = log(y[1]/y[0])/log(x[1]/x[0]);
+  }
+  else // No extrapolation
+    spl->der_lo = 0;
 
+  // High-end
+  if(spl->extrap_hi_type == ccl_f1d_extrap_const) {
+    spl->der_hi = 0;
+  } 
+  else if(spl->extrap_hi_type == ccl_f1d_extrap_linx_liny) {
+    spl->der_hi = (y[n-1]-y[n-2])/(x[n-1]-x[n-2]);
+  }
+  else if(spl->extrap_hi_type == ccl_f1d_extrap_linx_logy) {
+    if(y[n-1]*y[n-2]<=0)
+      *status=CCL_ERROR_SPLINE;
+    else
+      spl->der_hi = log(y[n-1]/y[n-2])/(x[n-1]-x[n-2]);
+  }
+  else if(spl->extrap_hi_type == ccl_f1d_extrap_logx_liny) {
+    if(x[n-1]*x[n-2]<=0)
+      *status=CCL_ERROR_SPLINE;
+    else
+      spl->der_hi = (y[n-1]-y[n-2])/log(x[n-1]/x[n-2]);
+  }
+  else if(spl->extrap_hi_type == ccl_f1d_extrap_logx_logy) {
+    if((y[n-1]*y[n-2]<=0) || (x[n-1]*x[n-2]<=0))
+      *status=CCL_ERROR_SPLINE;
+    else
+      spl->der_hi = log(y[n-1]/y[n-2])/log(x[n-1]/x[n-2]);
+  }
+  else // No extrapolation
+    spl->der_hi = 0;
   return spl;
 }
```

Comment:
Do we need to do anything with status above before we return here?

Reply:
Actually I've changed things. Now instead of changing the status, the extrapolator is set to "no extrapolation", so that, if extrapolation is ever needed, an error will be thrown.

Reply:
This is because often extrapolation won't be possible because of wrong edges, but it's also never needed anyway (it happens in some of the correlation function tests)

Reply:
Yikes this feels a bit too clever. It is not wrong, but I can see this causing an MCMC chain to fail right in the middle. I think maybe this is not the best idea.

Reply:
Humm, not sure why this is a problem (this is exactly the same behaviour that `ccl_correlation.c` had before).
1- Code initalizes this through CCL with some extrapolation option
2- CCL realizes that this extrapolation is not possible and defaults to "no extrapolation"
3- At some point the MCMC asks CCL to extrapolate, and CCL throws an error.

Why is this more dangerous than CLASS failing?

Reply:
Well if someone is asking for log interp but the bin limits are <=0, then that is an actual bug/problem. We should fail immediately instead of letting the error go along silently in the hopes we do not hit it. It is a matter if principle.

Reply:
These flags only apply to extrapolation, not interpolation. So we will fail only if extrapolation is ever needed. I think this is right, but I can change it to the previous option if you prefer.

Reply:
Yeah I think it should fail ASAP. People should have to think about this.

Reply:
OK, done.
To get back to your initial question: I don't think we need to do anything there, the status gets checked after initializing these structures in ccl_tracers and ccl_correlation.